### PR TITLE
Feature-free implementation of vectors and vector spaces

### PR DESCRIPTION
**Pull request by AdamNorberg**
Beginning of implementation for issue #45. Introduces the `IVectorSpace` concept, a few sample vector spaces, and the barest beginnings of the `Vector<IVectorSpace>` type itself - just constructors, accessors, and equality comparison. Actual vector math will be in later pull requests.

### CODE REVIEW

File: engine/calculus/Vectors/Vector.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+using System.Security.Policy;
+
+namespace FreedomOfFormFoundation.AnatomyEngine.Calculus.Vectors
+{
+    // A Vector represents a spatial vector in zero or more dimensions, existing in a specific vector space
+    // throughout its life. A Vector is immutable (operations will give new Vectors).
+    public struct Vector<TSpace> : IEquatable<Vector<TSpace>> where TSpace : IVectorSpace
+    {
+        public TSpace Space { get; }
+        private readonly Real[] _v;
```
Comment (Lathreas) [inline engine/calculus/Vectors/Vector.cs:31]: Using a variable-length array will make the Vector type have a few bytes of overhead, and since the array is allocated on the heap, will also lead to lower performance due to needing to allocate on the heap each time. Since we will be doing operations on huge arrays of vectors, this can become a big limitation on the use of these Vector types (and might even give rise to gigabytes of overhead in practice when displaying a full human). This will make it harder to use Vector for the more intensive applications. Since we need a lot of Vector operations, I think it will make sense to use explicit `x`,`y`, `z`, `w` variables. This probably requires implementing a separate Vector2, Vector3 and Vector4 class for each type, unfortunately, but I think the performance and memory gains will be worth it. We will likely not need larger size vector types for intensive applications, so those could use a variable-length array.

File: engine/calculus/Vectors/Vector.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+using System.Linq;
+using System.Security.Policy;
+
+namespace FreedomOfFormFoundation.AnatomyEngine.Calculus.Vectors
+{
+    // A Vector represents a spatial vector in zero or more dimensions, existing in a specific vector space
+    // throughout its life. A Vector is immutable (operations will give new Vectors).
+    public struct Vector<TSpace> : IEquatable<Vector<TSpace>> where TSpace : IVectorSpace
+    {
+        public TSpace Space { get; }
```
Comment (Lathreas) [inline engine/calculus/Vectors/Vector.cs:30]: Just to verify, if `TSpace` is an empty struct, does the compiler ensure there is no byte overhead? Or does this add a few bytes to each Vector instance? Again, this is very important to ensure we don't waste data when making large arrays.

File: engine/calculus/Vectors/Vector.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                int hash = 0x1B15D00D;  // arbitrary
+
+                foreach(Real r in _v)
+                {
+                    hash = hash << 1 ^ r.GetHashCode();
```
Comment (Lathreas) [inline engine/calculus/Vectors/Vector.cs:131]: This is clear to me now that this is to randomize the hash code a bit, but I think it might still help to put a comment in here that this arbitrary calculation is there to randomize the hash.

File: engine/calculus/Vectors/Vector.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        public TSpace Space { get; }
+        private readonly Real[] _v;
+
+        /// <summary>
+        /// Construct a Vector in the provided space, with the provided values. There must be exactly as many values
+        /// as there are dimensions in the vector space.
+        /// </summary>
+        /// <param name="space">Vector space for this vector.</param>
+        /// <param name="values">Values in this vector.</param>
+        public Vector(TSpace space, IEnumerable<Real> values)
```
Comment (Lathreas) [inline engine/calculus/Vectors/Vector.cs:39]: Do we need a reference to the TSpace for each instance of the vector? I think this should be enforced by the compiler rather than during runtime. Indeed this would then mean that there is no explicit coordinate system for the space for each Vector instance, but then again the programmer will know what space they are using in their calculations and keep a TSpace instance in the surrounding code. This would give higher performance and less byte overhead.

File: EngineTests/calculus/Vectors/ArbitraryVectorSpace.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace EngineTests.calculus.Vectors
+{
+    /// <summary>
+    /// ArbitraryVectorSpace is a vector space with a name and dimension, which is the same as
+    /// any other ArbitraryVectorSpace with the same name and dimension. It is not intended
+    /// for production use - it is for test use only. It lets us create arbitrary type-equal
+    /// and value-unequal vector spaces.
+    /// </summary>
+    public struct ArbitraryVectorSpace : IVectorSpace
```
Comment (Lathreas) [inline EngineTests/calculus/Vectors/ArbitraryVectorSpace.cs:11]: This struct would of course be used for debug purposes only, considering its huge overhead. Would it be an idea to make this only accessible to a `#define DEBUG` code branch? Or do you think there's value in keeping it for production use as well?

**Review by Lathreas — COMMENTED** [review]
Alright, here's my review! I think for this code to be used, it should not reduce performance too much. I think there are a few areas that can have better performance if rewritten, allowing this code to be used more extensively throughout the program. The current float-based code is already becoming too slow to handle (it takes a few seconds to generate a very simple bone, which is far from real-time; and some testing has shown that this is purely due to the arithmetics rather than due to array allocation), so I think it will be very important to ensure vector-based operations don't introduce additional overhead.

File: EngineTests/calculus/Vectors/ArbitraryVectorSpace.cs (hunk 5 above)
Comment (AdamNorberg) [reply EngineTests/calculus/Vectors/ArbitraryVectorSpace.cs:11]: The EngineTests project better not wind up compiled into a production release... 

I don't think there's a reason to explicitly block this from optimized builds, because I can see us wanting to do performance tests on how this behaves once the optimizer took a swing at it.

File: engine/calculus/Vectors/Vector.cs (hunk 1 above)
Comment (AdamNorberg) [reply engine/calculus/Vectors/Vector.cs:31]: C# doesn't actually have real fixed-length arrays. It has fixed buffers, which are explicitly designed for marshalling to unmanaged code and can only be used in a context declared as `unsafe`; fixed buffers can only be members of `structs` (never `class`es!) that are themselves `unsafe` (and, therefore, can only _exist_ in an `unsafe` context). `unsafe` would wind up covering the entire program and we'd be better off using a language designed for that use pattern, like C++ or Go.

C# arrays are therefore always heap-allocated and variable-length. The Vector class is either calling to an unmanaged library under the hood or it's literally using three separate named fields. Since it's designed for GPU-based execution, I'd bet it's an unmanaged library; there are costs for entering unamanaged/unsafe code from a managed context in C#. The JIT compiler does its best to keep all the safety guarantees of managed code while _exiting_ unmanaged contexts as reluctantly as possible, but I don't know how sophisticated it is at that.

I wouldn't be so sure that the heap-allocated array is as bad as it sounds. C# is optimized for this pattern. Like Java, nominally-heap-allocated objects with a trivially stack-limited lifespan are eligible for stack allocation if they're small and, if they aren't, go in a stack-frame-specific heap region that is invalidated as a whole at the end of the stack frame. If the amount of memory a function will allocate in stack-heap is easy to calculate at the start of the function, the JIT will usually allocate it in one go (and if it's small, it will just be part of the stack frame). This is the primary reason why swarms of tiny short-lived objects don't have as much performance impact in these languages as you might expect.

C# generics aren't C++ templates. C# generics, like Java generics, rely on _type erasure_ rather than compiled-in specialization. Every struct type of the same size gets the same implementation of a C# generic, and every heap type gets the same implementation (since all the pointers are the same size). There are pros and cons to this approach; I tend to regard it as mostly bad. It is impossible to access the parameterized type itself from a generic context in the absence of an object (or object reference) of that type; therefore, I cannot avoid using an "exemplar object" to implement context-aware vectors in C#. I'd use a different approach in Go or C++.

All that said: as absurd as it sounds, _don't assume this has worse performance than the Vector implementation until we run it._ Vector is specialized for math on GPUs, but this implementation isn't calling into any dispatch-to-GPU libraries and it's not obvious that we'll be able to do so in the near future. Therefore, this GPU-specialized library is being executed on a CPU. _Single-precision floating point math is slower than double-precision on a modern CPU, often much slower._ Modern CPUs typically do not perform single-precision floating point math in hardware, so doing single precision math involves an awful lot of converting back and forth between float formats as values exit registers and get stored back into RAM. Most CPUs _do_ have some support for this but it's slower when the `float`s are not 64-bit-aligned, as they naturally are not because they're 32-bit values (if you were going to spend the RAM on 64-bit alignment, you'd presumably be using `double`, right?).

All that said, this context-aware Vector type _will_ take more RAM for the out-of-line array alloc, the context object, the pointers to the out-of-line array and context, and the size of `double` vs `float`. Since the intent was never to use `float` for high precision mathematics in production, however, that last one is unavoidable, and then the other costs look half as bad by comparison.

I agree that this type needs to be written with an eye for performance. Among other things, I intend to wrap "throw an assertion if the vector space of this vector isn't the same as the vector space of that vector" logic in a function, which gets the `[Conditional("DEBUG")]` attribute so it is removed in debug builds, along with calls to it. Other safety checks can be treated similarly. 

 Anyway: short version is, I don't think the _structural_ performance flaws described here will necessarily impact performance, especially in optimized builds. I also don't think small constant factors will make a satisfying difference in performance of bone calculation; I _do_ think it's worth looking for ways to use multiple CPU cores during calculation (especially using high-level APIs that let the runtime decide the appropriate amount of parallelization to attempt) and aggressively searching for, and pruning, unnecessary work. 

That's true whether we use this vector implementation or not! For all my grousing about why the performance issues here may not exist, they also might exist and we'll just have to write three vector types (Vector2, Vector3, Vector4) and a truly obnoxious collection of matrix types (RowVector2, RowVector3, RowVector4, Matrix2x2, Matrix2x3, Matrix2x4, Matrix3x2, Matrix3x3, Matrix3x4, Matrix4x2, Matrix4x3, Matrix4x4), all of them by hand, to get acceptable performance. Although at that point I think there's an argument to be made that the constraints of C#'s type model are too severe for the type of high performance math we want to do, and start over in C++. (I would advocate for Go, but Go isn't necessarily better than C# here as far as annoying repetition goes.) But we also definitely need to give the raytracer an optimization pass. I think we need to start by writing some benchmark routines so we have something to measure, and then run them with CPU sampling to find out what the exact hotspots are - it sounds like you've already done this to some extent?

File: engine/calculus/Vectors/Vector.cs (hunk 2 above)
Comment (AdamNorberg) [reply engine/calculus/Vectors/Vector.cs:30]: `TSpace` may be the smallest struct, which is zero bytes of data plus a type header. Not zero, unfortunately, since C# doesn't compile a separate specialization for every set of type arguments to a generic.

File: engine/calculus/Vectors/Vector.cs (hunk 3 above)
Comment (AdamNorberg) [reply engine/calculus/Vectors/Vector.cs:131]: Is such a comment valuable? It seems intuitive to me that any deterministic but messy math operation in any GetHashCode implementation can be inferred to be for the purpose of calculating a hash...

File: engine/calculus/Vectors/Vector.cs (hunk 4 above)
Comment (AdamNorberg) [reply engine/calculus/Vectors/Vector.cs:39]: Unfortunately, unlike C++, in C# there is no way to get type-specific behavior out of a type parameter without an object reference at some point, due to its type erasure model. C# generics can only refer to fields or methods on a parameterized type that are known to exist because of constraints on that type (the `where` clause in the declaration that produced the type parameter). The only type constraints that define behavior instead of memory lay out are "default constructible", "is subclass of", and "implements interface". The latter two only offer polymorphic behavior for instances.

In practice, the JIT is likely to "compile down" dynamic calls into static calls when it has enough type information to do so - and, from there, can go from static calls into constants when it's calling functions that it can prove have no side effects - but none of this allows us to get rid of the TSpace instance, annoyingly.

**Comment by AdamNorberg** [comment]
TL;DR: C# generics do not offer substantial code generation, unlike C++ templates, so concrete fields have to hang around just to hold type information that does not otherwise exist at runtime.

**Comment by AdamNorberg** [comment]
Also, all this is stuff I _wanted_ to do, but can't do in pure C#. Using C++/CLR (AKA, writing C++ for .NET) barely helps; we could use templates for all of this, but we'd have to handwrite a class (or at least write a macro to write a class...) for _every specialization_ that needs to be visible in C# because C# generics cannot pass parameterized type information down to C++ templates at compile time, so no template specialization can be made for a parameterized type.

That might still be the least bad option, for what that's worth, since we could still write an optimized C++ class designed for stack allocation, compose it into its .NET-ready wrapper types (observed by C# as `struct`s), and get the behavior we predict is likely to be highest-performance while minimizing code repetition (we have to write a bunch of wrappers but that's way less work than rewriting all the _bodies_ too - enough less work that we can realistically write context-specific wrappers).

File: EngineTests/calculus/Vectors/ArbitraryVectorSpace.cs (hunk 5 above)
Comment (Lathreas) [reply EngineTests/calculus/Vectors/ArbitraryVectorSpace.cs:11]: Oh oops, didn't check the folder it was in. You're right, it's already in debug, my bad! :P

File: engine/calculus/Vectors/Vector.cs (hunk 3 above)
Comment (Lathreas) [reply engine/calculus/Vectors/Vector.cs:131]: That's fair. I was thinking a bit about my former self who _wasn't_ 100% sure that's what it was for, but you're quite possibly right that it's superfluous. That said, I like comments beside arithmetics :P.

**Comment by Lathreas** [comment]
Hmmm, yeah, thanks for the in-depth response! I definitely agree, and the limit of C#'s generics was something I have been afraid of for quite a while now, to be fair. I recognize the struggles you've described, and although indeed this might very well be the highest-performance implementation we can get to in C#, I do fear it might have long-term consequences to the performance of the entire application if we keep holding on to it for too long. The resource requirements are only really expected to increase, and indeed rewriting this Vector type using C++/CLR is not really an attractive option due to repetitive glue code and the inevitable marshaling overhead that we will get.

> Also, all this is stuff I wanted to do, but can't do in pure C#.

Oh, yeah, I feel your pain :P (see: the horrible ContinuousMap<TIn, TOut> which I had hoped to make fully generic but instead had to implement specific versions of, kind of removing the purpose of the generics :P).

I'm fine using a less-optimized Vector struct implementation for the sake of debugging and development, but I do believe that for any production work we need to squeeze out all of the performance we can get; especially on the lower level. Although performance is very important, I'm also particularly afraid of memory overhead, which will become severely limiting if we want to fully represent anatomy at the sub-millimeter level. I have already made a few Blender surfaces using careful sculpting and memory optimization techniques at millimeter-level precision, and despite using no-overhead buffers and 32bit precision, the arrays themselves easily surpass 6 GB of memory (for merely the face and hair follicles) despite being topologically optimized (i.e. using the 'decimate' modifier to remove unnecessary detail). Due to the nature of the program, we will need to be using topologically _unoptimized_ surfaces, since the program cannot possibly predict the level of detail a particular spot can have. Of course, quadtrees come into mind for optimizing that structure, but that'll still be nowhere near as memory efficient as a decimate modifier. These memory costs should be taken into account, so for any large array storage we should be careful not to store unnecessary overhead. For small stuff, such as representing the broad shapes and curves, the memory overhead will probably not impact too much.

For performance, I haven't done any 'real' testing in the sense that I used proper measurement tools, but I did do a few simple timing checks. It appears a lot of the delay is in the calculation of the intersection points, although I haven't yet determined whether this is specifically due to the arithmetic or due to e.g. method call overhead (I expect the former though). Indeed we should opt for better performance measurements than that, but the order of magnitude should remain the same either way.

I'll definitely do a few general arithmetic optimization passes in the near future, and that would indeed be generally applicable regardless of the language. I think there are indeed huge performance gains to be gotten from simply skipping code that isn't necessary, or by trying to simplify some of the mathematics. Doing parallelization will require us to rethink how the code is executed; using iterators will make the process inherently serial whereas the code can easily be parallelized using a preallocated array.

That said, I'm not too comfortable doing a lot of architecture-dependent optimization work if we aren't sure we will stick to the language at hand unless it's a very obvious and easy change (such as changing a type or adding a compiler hint), at risk of doing too much unnecessary work (remember that we only really spend a few hours each week on this currently and development time is quite valuable).

So all in all, if indeed C# is simply too limiting for the work we intend to do, I do think we should plan ahead a time to move to a lower-level language for the engine code if we are to make the library production-ready, either now or after a first draft is ready. I think this might have multiple benefits, also to make it more of a general-use library beyond just C# (since C++ is easier to integrate into many languages from my experience). Furthermore, from my experience with .NET and massively parallel processing, I believe I will also be able to get more GPU-performance using carefully written libraries if we use C++/OpenCL or CUDA for the important bits instead of using C#'s methods, since I'll have more control over the CPU-GPU communication overhead. So far, I've found hand-written communication code to have superior performance over auto-generated GPU parallelization because the compiler cannot predict whether arrays will be used on the GPU or on the CPU, so it copies the final result back into memory each time it is accessed (either in advance or just-in-time, so it's not all wasted, but for example between class calls it will often copy the result back onto memory).

**Comment by AdamNorberg** [comment]
Not practical. Using GLM#.
